File: javascript/hello_world.js
#!/usr/bin/env node
var Print = function (str) {
    console.log(str);
};

function HelloWorld () {

}

HelloWorld.prototype.product = function() {
    var product = 1;
    var _length = arguments.length;
    for (var i = 0; i < arguments.length; i++) {
        product = product* arguments[i];
    }
    return product;
};
HelloWorld.prototype.range = function(start,stop) {
    var range = Array();
    for (var i = start; i<stop; i++) {
        range.push(i); // Javascript >= 1.7 has a 'yield' statement
    }
    return range;
}

HelloWorld.prototype.main = function() {
    // stdout
    Print("Hello world");

    // basic math
    Print("// basic math");
    var _sum = 0;
    _sum = 2.0 + 2.0;
    Print(_sum);
    var sqrt = Math.sqrt(_sum);
    Print(sqrt);

    // for and while loops; nested arrays; functions
    Print("// loops; nested arrays; functions");
    var n = 10;
    var nsums = Array(n);
    var totalsum = 0;
    var i = 0;
    while (i < n) {
        nsums[i] = Array(n);
        for (var j = n-1; j >= 0; j--) {
            nsums[i][j] = this.product(i,j);
            totalsum += nsums[i][j];
        }
        i+=1;
    }

    Print("// if-conditionals");
    if (totalsum===2025) {
        Print("totalsum*5: " + totalsum*5);
    } else if (totalsum===0) {
        Print("totalsum is zero?");
    } else {
        print("totalsum is off");
    }

    // function overload (what is this called again?)
    Print("// function overload"); // not really
    Print(this.product(2,3,4));

    // input coercion
    Print("// input coercion");
    var ten = parseInt("10"); 
    var tenpointoh = parseFloat("10.0");

    // try/except blocks
    Print("// try/catch blocks");
    var doublestr = "x20.2";
    try {
        var twentypointtwo = parseFloat(doublestr);
        if (twentypointtwo === NaN) {
            throw ("'" + twentypointtwo + "' is not a double");
        }
    } catch(err) {
        Print("err: " + doublestr + " is not a double");
        Print(err);
    } finally {
        doublestr = doublestr.substring(1,-1); // TODO
        var twentypointtwo = parseFloat(doublestr);
    }

    ////
    // Arraylists
    Print("// ArrayLists");
    var intlist = this.range(0,5);
    Print(intlist.length);
    var _ = intlist.pop(0);
    Print(intlist.length);

    // traditional iteration
    var _length = intlist.length;
    for (var i=0; i < _length; i++) {
        Print(intlist[i]);
    }

    // iteration with downcasting
    var _length = intlist.length;
    for (var i=0; i < _length; i++) {
        Print(intlist[i]);
    }

    // string join
    Print(intlist.join("\n"));

    ////
    // HashMaps
    Print("// Maps");
    var map = {};
    map['one'] = 1;
    map['two'] = 2;
    map['three'] = 3;
    map['four'] = 4;
    Print("size = " + map.length); // TODO: uhhh

    var key = "one";
    if ( map.hasOwnProperty(key) ) {
        Print(key + ": " + map[key]);
    }

    var searchvalue = "four";
    var foundvalue = false;
    for (key in map) {
        if (map[key] === searchvalue) {
            Print("contains the value '" + searchvalue + "'");
            foundvalue = true;
        }
    }
    if (!foundvalue) {
        Print("does not contain the value '" + searchvalue + "'");
    }

    for (key in map) {
        Print(key+"="+map[key]);
    }

    ////
    // Collections, sorting
    Print("// Collections.sort");
    var keys = [];
    for (key in map) {
        keys.push(key);
    }
    keys = keys.sort();
    var _length = keys.length;
    for (var i = 0; i < _length; i++) {
        var key = keys[i];
        Print(key + "=" + map[key]);
    }

    ////
    // OrderedMap
    Print("// OrderedMap");
    var mapkeys = [];
    var sortedmap = {};
    var pushKey = function(obj, key, value) {
        mapkeys.push(key);
        sortedmap[key] = value;
    }
    pushKey(sortedmap, "one", 1);
    pushKey(sortedmap, "two", 2);
    pushKey(sortedmap, "three", 3);
    pushKey(sortedmap, "four", 4);
    var _keylength = mapkeys.length;
    for (var i = 0; i < _keylength; i++) {
        var _key = keys[i];
        Print(_key + "=" + sortedmap[_key]);
    }

    ////
    // invert dict
    pushKey(sortedmap, "uno", 1);
    invertedmap = {};
    for (var i = 0; i < _keylength; i++) {
        var _key = keys[i];
        var _value = sortedmap[_key];
        var _currentvalue = invertedmap[_key];
        if (_currentvalue !== undefined && _currentvalue !== _key) { // TODO
            Print("duplicate key: '" + _value + "' [" + _key + 
                    "] {{" + _currentvalue + "}}");
        } else {
            invertedmap[_value] = _key;
        }
    }
    for (key in invertedmap) {
        Print(key + "=" + invertedmap[key]);
    }

};

var hw = new HelloWorld;
hw.main();

File: ruby/hello_world.rb
class HelloWorld
    def product(*args)
        return args.reduce(:*)
    end

    def main
        # stdout
        puts "Hello World";

        # basic math
        puts "//basic math"
        sum = 0
        sum = 2.0 + 2.0
        puts sum
        sqrt = Math.sqrt(sum)
        puts sqrt

        # for and while loops; nested arrays; functions
        puts "// loops; nested arrays; functions"
        n = 10
        nsums = []
        totalsum = 0
        i = 0
        while (i < n)
            nsums.push [nil,]*n
            for j in (0..n-1).reverse_each
                nsums[i][j] = product(i,j)
                totalsum += nsums[i][j]
            end
            i+=1
        end

        # if-conditionals
        puts "// if-conditionals"
        if totalsum == 2025
            puts "totalsum*5: #{totalsum*5}"
        elsif totalsum == 0
            puts "totalsum is zero?"
        else
            puts "totalsum is off"
        end

        # function overload (what is this called again?)
        puts "//function overload"
        puts product(2,3,4) # TODO

        # input coercion
        puts "// input coercion"
        ten = Integer("10")
        tenpointoh = Float("10.0")

        # begin/rescue blocks
        puts "// begin/rescue blocks"
        doublestr = "x20.2"
        begin
            twentypointtwo = Float(doublestr)
        rescue
            puts "err: '#{doublestr}' is not a double"
            puts "#{$!}"
        ensure
            doublestr = doublestr[1..-1]
            twentypointtwo = Float(doublestr)
        end

        ##
        # Arraylists
        puts "// ArrayLists"
        intlist = Array(0..4)
        puts intlist.length
        intlist.pop 0
        puts intlist.length

        # traditional iteration
        intlist.each do |nextint|
            puts nextint
        end

        # iteration with downcasting
        intlist.each do |nextint|
            puts nextint
        end

        # ruby iteration
        puts intlist.join("\n")

        ##
        # HashMaps
        puts "// Maps"
        map = Hash.new
        map['one'] = 1
        map['two'] = 2
        map['three'] = 3
        map['four'] = 4
        puts "size = #{map.length}"

        key = "one"
        if map.key? key # also has_key?
            puts "#{key}: #{map[key]}"
        end

        searchvalue = "four"
        if map.value? searchvalue # also has_value?
            puts "contains the value '#{searchvalue}'"
        else
            puts "does not contain the value '#{searchvalue}'"
        end

        map.each do |k,v|
            puts "#{k}=#{v}"
        end

        ##
        # Collections, sorting
        #
        puts "// Collections.sort"
        map.keys.each do |k|
            puts "#{k}: #{map[k]}"
        end

        ##
        # OrderedDict
        # ruby >= 1.9 maintains Hash insertion order
        puts " // OrderedDict"
        sortedmap = Hash.new
        sortedmap['one'] = 1
        sortedmap['two'] = 2
        sortedmap['three'] = 3
        sortedmap['four'] = 4
        puts "size = #{sortedmap.length}"
        sortedmap.each do |k,v|
            puts "#{k}=#{v}"
        end

        ##
        # invert dict
        sortedmap['uno'] = 1
        invertedmap = Hash.new
        sortedmap.each do |k,v|
            _currentvalue = invertedmap[v] # XXX: "default" = nil (sentinel)
            if _currentvalue != nil
                puts "duplicate key: '#{v}' [#{k}] {{#{_currentvalue}}}"
            else
                invertedmap[v] = k
            end
        end
        invertedmap.each do |k,v|
            puts "#{k}=#{v}"
        end
    end
end 
        

if __FILE__ == $0
    hw = HelloWorld.new
    hw.main
end
